IPC Organ Printing

CASEYGUNGEON opened the pull request:
Version 1: It Works on my End!

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Apparently Exosuit Fabricators could build IPC organs before, but now they can't. Except NOW now they can. Cheap early-game tech.

## How This Contributes To The Skyrat Roleplay Experience

<!-- Please add a short description of why you think these changes would benefit the game and the roleplay atmosphere of the server. If you can't justify it in words, it might not be worth adding. -->

IPCs can get missing organs put back in them, organics can get their organs replaced, (tested; does in fact let organics interface with APCs with only organ replacement) and maybe IPCs can be made from scratch.


## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Early-game tech allows Exosuit Fabs to print IPC organs.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

CASEYGUNGEON commented:
Oh, whoops. I didn't know updating my fork would mess with this pull request. I guess I need to reverse that commit or something.

CASEYGUNGEON commented:
Sorry for the mess. First time.

CASEYGUNGEON commented:
Relatively minor thing, but i noticed the pre-req isn't actually the one i wrote down in the PR. it's the one that starts already done that's called like Basic Medical Equipment. I don't really think that's an issue so I didn't push anything else.

GoldenAlpharex commented on code/modules/vehicles/mecha/mech_fabricator.dm:
```diff
@@ -191,6 +194,13 @@
 
+		// SKYRAT EDIT START - ipc_printing
+
+		else if(findtext(built_item, "robot_ipc") || findtext(built_item,"power_cord"))
```
I got a feeling that's probably not the best way to do this, have you tried using `ispath()` like the other ones?

GoldenAlpharex commented on modular_skyrat/modules/ipc_printing/code/ipc_designs.dm:
```diff
@@ -0,0 +1,55 @@
+	name = "IPC Part"
+	build_type = MECHFAB
+	departmental_flags = DEPARTMENTAL_FLAG_SCIENCE
+	materials = list(/datum/material/iron=1000)
```
```suggestion
	materials = list(/datum/material/iron = 1000)
```

GoldenAlpharex commented on modular_skyrat/modules/ipc_printing/code/ipc_designs.dm:
```diff
@@ -0,0 +1,55 @@
+	build_type = MECHFAB
+	departmental_flags = DEPARTMENTAL_FLAG_SCIENCE
+	materials = list(/datum/material/iron=1000)
+	construction_time = 50
```
```suggestion
	construction_time = 5 SECONDS
```

GoldenAlpharex commented on modular_skyrat/modules/ipc_printing/code/ipc_printing_techweb.dm:
```diff
@@ -0,0 +1,16 @@
+	id = "ipc_parts"
+	display_name = "IPC Organs"
+	description = "Allows Exosuit Fabricators to build IPC organs."
+	prereq_ids = list("basic_medical")
```
That means it'll be available to be researched by default. I'd probably move that to be after something as part of the robotics research tree.

CASEYGUNGEON replied:
yeah, makes sense. i'd originally meant it to have the same pre-req as cyber organs but read the labels wrong at the time. i'll take another look at them.

GoldenAlpharex reviewed (changes requested):
Good change, only a few things to look through.

CASEYGUNGEON commented:
not sure why it's failing trying to make/destroy shadow bodyparts. i guess i'll try and take a look at that later today.

Zergspower commented:
Does this add their sprites in? Currently half of them dont have them

CASEYGUNGEON commented:
Just checked. Only the reagent processing unit doesn't have a sprite. guess i just never counted when i tried to print all of them.
this is the other 7:
[https://media.discordapp.net/attachments/503781607769899039/963130410940903454/unknown.png](url)

guess i gotta draw up a sprite or something.

CASEYGUNGEON commented:
Good news! the reason why the liver didn't have a sprite is because it was looking for `liver-c` instead of `liver-ipc`. i guess technically that's a change so i added it to the readme and put a comment next to where I fixed it. All 8 organs have sprites.

CASEYGUNGEON commented:
reorganized my fork, didn't know it would auto-close the PR. not sure about clogging everything with a second PR, but I can't reopen this so i can do that if it's not too weird. shouldn't have to reorganize again. (this PR was pulling from my main branch. wanted to be able to fetch upstream and try other changes and not have this stuff be in my main)
